dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordModelBuilder.cs

```diff
@@ -0,0 +1,493 @@
+                    }
+
+                    break;
+
+                default:
+                    throw new ArgumentException($"Unknown type '{definitionProperty.GetType().FullName}' in vector store record definition.");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Extension hook for connectors to be able to customize the model.
+    /// </summary>
+    protected virtual void Customize()
+    {
+    }
+
+    /// <summary>
+    /// Finalizes the model after all properties have been processed, validating the model and applying some final transformations.
+    /// </summary>
+    protected virtual void Finalize(Type clrType)
+    {
+        if (!this.Options.UsesExternalSerializer && clrType.GetConstructor(Type.EmptyTypes) is null)
+        {
+            throw new NotSupportedException($"Type '{clrType.Name}' must have a parameterless constructor.");
+        }
+
+        if (!this.Options.SupportsMultipleKeys && this.KeyProperties.Count > 1)
+        {
+            throw new NotSupportedException($"Multiple key properties found on type '{clrType.Name}' or the provided {nameof(VectorStoreRecordDefinition)}.");
+        }
+
+        if (this.KeyProperties.Count == 0)
+        {
+            throw new NotSupportedException($"No key property found on type '{clrType.Name}' or the provided {nameof(VectorStoreRecordDefinition)}.");
+        }
+
+        if (this.Options.ReservedKeyStorageName is not null)
+        {
+            // Single key property validated in the constructor
+            this.KeyProperties.Single().StorageName = this.Options.ReservedKeyStorageName;
+        }
+
+        if (this.Options.RequiresAtLeastOneVector && this.VectorProperties.Count == 0)
+        {
+            throw new NotSupportedException($"No vector property found on type '{clrType.Name}' or the provided {nameof(VectorStoreRecordDefinition)} while at least one is required.");
+        }
+
+        if (!this.Options.SupportsMultipleVectors && this.VectorProperties.Count > 1)
+        {
+            throw new NotSupportedException($"Multiple vector properties found on type '{clrType.Name}' or the provided {nameof(VectorStoreRecordDefinition)} while only one is supported.");
+        }
+
+        var storageNameMap = new Dictionary<string, VectorStoreRecordPropertyModel>();
+
+        foreach (var property in this.PropertyMap.Values)
+        {
+            this.FinalizeProperty(property);
+
+            if (storageNameMap.TryGetValue(property.StorageName, out var otherproperty))
+            {
+                throw new InvalidOperationException($"Property '{property.ModelName}' is being mapped to storage name '{property.StorageName}', but property '{otherproperty.ModelName}' is already mapped to the same storage name.");
+            }
+
+            storageNameMap[property.StorageName] = property;
+        }
+    }
+
+    /// <summary>
+    /// Finalizes a single property, performing validation on it.ssss
```

Comment:
The `ssss` reminds me of the Harry Potter movie where it turned out that Harry can speak Parseltongue.

```suggestion
    /// Finalizes a single property, performing validation on it.
```